shaping/input.go

```diff
@@ -33,3 +35,85 @@ type Input struct {
 	// Language is an identifier for the language of the text.
 	Language language.Language
 }
+
+// SplitByFontGlyphs split the runes from 'input' to several items, sharing the same
+// characteristics as 'input', expected for the `Face` which is set to
+// the first font among 'availableFonts' providing support for all the runes
+// in the item.
+// Runes supported by no fonts are mapped to the first element of 'availableFonts', which
+// must not be empty.
+// The 'Face' field of 'input' is ignored: only 'availableFaces' are consulted.
+// Rune coverage is obtained by calling the NominalGlyph() method of each font.
+func SplitByFontGlyphs(input Input, availableFaces []font.Face) []Input {
+	var splitInputs []Input
+	currentInput := input
+	for i := input.RunStart; i < input.RunEnd; i++ {
+		r := input.Text[i]
+		if currentInput.Face != nil && ignoreFaceChange(r) {
+			// add the rune to the current input
+			continue
+		}
+
+		// select the first font supporting r
+		selectedFace := availableFaces[0]
+		for _, f := range availableFaces {
+			if _, has := f.NominalGlyph(r); has {
+				selectedFace = f
+				break
+			}
+		}
+
+		if currentInput.Face == selectedFace {
+			// add the rune to the current input
+			continue
+		}
+
+		// new face needed
+
+		if i != input.RunStart {
+			// close the current input ...
+			currentInput.RunEnd = i
+			// ... add it to the output ...
+			splitInputs = append(splitInputs, currentInput)
+		}
+
+		// ... and create a new one
+		currentInput = input
+		currentInput.RunStart = i
+		currentInput.Face = selectedFace
+	}
+
+	// close and add the last input
+	currentInput.RunEnd = input.RunEnd
+	splitInputs = append(splitInputs, currentInput)
+	return splitInputs
+}
+
+// ignoreFaceChange returns `true` is the given rune should not trigger
+// a change of font.
+//
+// We don't want space characters to affect font selection; in general,
+// it's always wrong to select a font just to render a space.
+// We assume that all fonts have the ASCII space, and for other space
+// characters if they don't, HarfBuzz will compatibility-decompose them
+// to ASCII space...
+//
+// We don't want to change fonts just for variation selectors.
+//
+// Finally, don't change fonts for line or paragraph separators.
+//
+// The rationale is taken from pango : see bugs
+// https://bugzilla.gnome.org/show_bug.cgi?id=355987
+// https://bugzilla.gnome.org/show_bug.cgi?id=701652
+// https://bugzilla.gnome.org/show_bug.cgi?id=781123
+// for more details.
+func ignoreFaceChange(r rune) bool {
+	return unicode.Is(unicode.Cc, r) || // control
+		unicode.Is(unicode.Cf, r) || // format
+		unicode.Is(unicode.Cs, r) || // surrogate
+		unicode.Is(unicode.Zl, r) || // line separator
+		unicode.Is(unicode.Zp, r) || // paragraph separator
+		(unicode.Is(unicode.Zs, r) && r != '\u1680') || // space separator != OGHAM SPACE MARK
+		('\U000e0100' <= r && r <= '\U000e01ef') ||
+		('\ufe00' <= r && r <= '\ufe0f') // variation selectors
+}
```

Comment:
I think a better source for these code points would be https://github.com/harfbuzz/harfbuzz/blob/067f90a82087a5feb1cf76625c54c60a10946fe1/src/hb-unicode.hh#L131-L200. The code from Pango will not enable fallback for U+06DD (a long outstanding bug in Pango) because Unicode unhelpfully categorize it as format character, even though it always have a graphical representation.

Reply:
Thanks for your input !

To be certain, you suggest to ignore the code points in `is_default_ignorable`, as well as in the categories : 

- Separator, {Space, Line, Paragraph}
- Other, {Control, Surrogate}

(the correct code points in category Other, Format being handled by `is_default_ignorable`) 

Did I get this right ?

Reply:
Yes, sounds right.